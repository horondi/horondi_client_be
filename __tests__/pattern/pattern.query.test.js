const { gql } = require('@apollo/client');
/* eslint-disable no-undef */
const { PATTERN_NOT_FOUND } = require('../../error-messages/pattern.messages');
const { setupApp } = require('../helper-functions');
jest.mock('../../modules/upload/upload.service');

const {
  patternDoesNotExistId,
  skip,
  limit,
  wrongLimit,
  wrongSkip,
  user,
  languageTypeName,
  imageTypeName,
  queryPatternToAdd,
  testImages,
} = require('./pattern.variables');

let patternId;
let operations;

describe('Pattern queries', () => {
  beforeAll(async () => {
    operations = await setupApp();
  });

  beforeAll(async () => {
    const res = await operations
      .mutate({
        mutation: gql`
          mutation($pattern: PatternInput!) {
            addPattern(pattern: $pattern, image: []) {
              ... on Pattern {
                _id
              }
              ... on Error {
                message
                statusCode
              }
            }
          }
        `,

        variables: { pattern: queryPatternToAdd },
      })
      .catch(e => e);
    patternId = res.data.addPattern._id;
  });

  describe('tests for all patterns and one pattern', () => {
    test('Should receive all patterns', async () => {
      const res = await operations
        .query({
          query: gql`
            query {
              getAllPatterns {
                items {
                  name {
                    lang
                    value
                  }
                  description {
                    lang
                    value
                  }
                  images {
                    large
                    medium
                    small
                    thumbnail
                  }
                  material
                  handmade
                  available
                }
              }
            }
          `,
        })
        .catch(e => e);

      expect(res.data.getAllPatterns).toMatchSnapshot();
      expect(res.data.getAllPatterns).toBeDefined();
      expect(res.data.getAllPatterns.items).toContainEqual({
        name: queryPatternToAdd.name.map(item => ({
          ...languageTypeName,
          ...item,
        })),
        description: queryPatternToAdd.description.map(item => ({
          ...languageTypeName,
          ...item,
        })),
        images: { ...imageTypeName },
        material: queryPatternToAdd.material,
        handmade: false,
        available: true,
      });
    });
    test('Should receive one pattern', async () => {
      const res = await operations
        .query({
          query: gql`
            query($id: ID!) {
              getPatternById(id: $id) {
                ... on Pattern {
                  name {
                    lang
                    value
                  }
                  description {
                    lang
                    value
                  }
                  images {
                    large
                    medium
                    small
                    thumbnail
                  }
                  material
                  handmade
                  available
                }
              }
            }
          `,
          variables: { id: patternId },
        })
        .catch(e => e);

      expect(res.data.getPatternById).toMatchSnapshot();
      expect(res.data.getPatternById).toBeDefined();
      expect(res.data.getPatternById).toHaveProperty('images', testImages);
      expect(res.data.getPatternById).toHaveProperty(
        'name',
        queryPatternToAdd.name.map(item => ({
          ...languageTypeName,
          ...item,
        }))
      );
      expect(res.data.getPatternById).toHaveProperty('images', {
        ...imageTypeName,
        ...queryPatternToAdd.images,
      });
      expect(res.data.getPatternById).toHaveProperty(
        'description',
        ...queryPatternToAdd.description.map(item => ({
          ...languageTypeName,
          ...item,
        }))
      );
      expect(res.data.getPatternById.description).toBeInstanceOf(Array);
      expect(res.data.getPatternById).toHaveProperty(
        'material',
        queryPatternToAdd.material
      );
      expect(res.data.getPatternById).toHaveProperty(
        'handmade',
        queryPatternToAdd.handmade
      );
      expect(res.data.getPatternById).toHaveProperty(
        'available',
        queryPatternToAdd.available
      );
    });
    test('request not existing pattern should throw error', async () => {
      const res = await operations
        .query({
          query: gql`
            query($id: ID!) {
              getPatternById(id: $id) {
                ... on Pattern {
                  name {
                    lang
                    value
                  }
                  description {
                    lang
                    value
                  }
                  images {
                    large
                    medium
                    small
                    thumbnail
                  }
                  material
                  handmade
                  available
                }
                ... on Error {
                  statusCode
                  message
                }
              }
            }
          `,
          variables: { id: patternDoesNotExistId },
        })
        .catch(e => e);

      expect(res.data.getPatternById).toBeDefined();
      expect(res.data.getPatternById).toHaveProperty('statusCode', 404);
      expect(res.data.getPatternById).toHaveProperty(
        'message',
        PATTERN_NOT_FOUND
      );
    });

    test('pattern pagination test', async () => {
      const res = await operations
        .query({
          variables: { skip, limit },
          query: gql`
            query($skip: Int, $limit: Int) {
              getAllPatterns(skip: $skip, limit: $limit) {
                items {
                  name {
                    lang
                    value
                  }
                  description {
                    lang
                    value
                  }
                  images {
                    large
                    medium
                    small
                    thumbnail
                  }
                  material
                  handmade
                  available
                }
                count
              }
            }
          `,
        })
        .catch(e => e);

      expect(res.data.getAllPatterns).toMatchSnapshot();
      expect(res.data.getAllPatterns.items).toHaveLength(1);
      expect(res.data.getAllPatterns.count).toEqual(1);
    });
    test('Expect negative values', async () => {
      const res = await operations
        .query({
          variables: { skip: wrongLimit, limit: wrongSkip },
          query: gql`
            query($skip: Int, $limit: Int) {
              getAllPatterns(skip: $skip, limit: $limit) {
                items {
                  name {
                    lang
                    value
                  }
                  description {
                    lang
                    value
                  }
                  images {
                    large
                    medium
                    small
                    thumbnail
                  }
                  material
                  handmade
                  available
                }
                count
              }
            }
          `,
        })
        .catch(e => e);
      expect(res.errors[0].message).toEqual(
        'Skip value must be non-negative, but received: -5'
      );
    });
  });
  afterAll(async () => {
    await operations
      .mutate({
        mutation: gql`
          mutation($id: ID!) {
            deletePattern(id: $id) {
              ... on Pattern {
                _id
              }
              ... on Error {
                statusCode
                message
              }
            }
          }
        `,
        variables: { id: patternId },
      })
      .catch(e => e);
    expect(res.errors[0].message).toEqual(
      'Skip value must be non-negative, but received: -5'
    );
  });
});
