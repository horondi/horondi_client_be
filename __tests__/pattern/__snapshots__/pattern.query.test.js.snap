// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pattern queries tests for all patterns and one pattern Should receive all patterns 1`] = `
Object {
  "items": Array [
    Object {
      "available": true,
      "description": Array [
        Object {
          "lang": "uk",
          "value": "test value",
        },
        Object {
          "lang": "en",
          "value": "test value",
        },
      ],
      "handmade": false,
      "images": Object {
        "large": "large_0_test-file",
        "medium": "medium_0_test-file",
        "small": "small_0_test-file",
        "thumbnail": "thumbnail_0_test-file",
      },
      "material": "test",
      "name": Array [
        Object {
          "lang": "uk",
          "value": "test value",
        },
        Object {
          "lang": "en",
          "value": "test value",
        },
      ],
    },
  ],
}
`;

exports[`Pattern queries tests for all patterns and one pattern Should receive one pattern 1`] = `
Object {
  "available": true,
  "description": Array [
    Object {
      "lang": "uk",
      "value": "test value",
    },
    Object {
      "lang": "en",
      "value": "test value",
    },
  ],
  "handmade": false,
  "images": Object {
    "large": "large_0_test-file",
    "medium": "medium_0_test-file",
    "small": "small_0_test-file",
    "thumbnail": "thumbnail_0_test-file",
  },
  "material": "test",
  "name": Array [
    Object {
      "lang": "uk",
      "value": "test value",
    },
    Object {
      "lang": "en",
      "value": "test value",
    },
  ],
}
`;

exports[`Pattern queries tests for all patterns and one pattern pattern pagination test 1`] = `
Object {
  "count": 1,
  "items": Array [
    Object {
      "available": true,
      "description": Array [
        Object {
          "lang": "uk",
          "value": "test value",
        },
        Object {
          "lang": "en",
          "value": "test value",
        },
      ],
      "handmade": false,
      "images": Object {
        "large": "large_0_test-file",
        "medium": "medium_0_test-file",
        "small": "small_0_test-file",
        "thumbnail": "thumbnail_0_test-file",
      },
      "material": "test",
      "name": Array [
        Object {
          "lang": "uk",
          "value": "test value",
        },
        Object {
          "lang": "en",
          "value": "test value",
        },
      ],
    },
  ],
}
`;
